grepcidr.c

```diff
@@ -62,13 +62,46 @@ static unsigned int counting = 0;		/* when non-zero, counts matches */
 static int invert = 0;				/* flag for inverted mode */
 static int anchor = 0;				/* anchor matches at beginning of line */
 static int nonames = 0;				/* don't show filenames */
+static int onlynames = 0;			/* only show filenames */
 static int nmatch = 0;				/* count of matches for exit code */
+static int fmatch = 0;				/* count of matches for current file */
 static int igbadpat = 0;			/* ignore bad patterns */
 static int sloppy = 0;				/* don't complain about sloppy CIDR */
 static int cidrsearch = 0;			/* parse and match CIDR in haystack */
 static int didrsearch = 0;			/* match CIDR if overlaps with haystack */
 static int quick = 0;				/* quick match, ignore v4 with dots before or after */
 
+/* Global scanning state for scan_block */
```

Comment:
Globalizing the `scan_state` variable of function `scan_block` was necessary so that the state value could communicated back to the `scan_read` function's `while` loop.  This allows the scan engine to tell `scan_read` that there is no reason to continue scanning the file.  This in turn means that the `-l` flag is sped up by aborting the scan as soon as the first match is found: the filename is printed, and scanning moves on to the next file.